ocaml/xapi-cli-server/cli_operations.ml

```diff
@@ -4310,24 +4310,42 @@ let vm_migrate printer rpc session_id params =
       Client.Session.login_with_password remote_rpc username password "1.3"
         Constants.xapi_user_agent
     in
+    let remote f = f ~rpc:remote_rpc ~session_id:remote_session in
     finally
       (fun () ->
-        let host, host_record =
-          let all = Client.Host.get_all_records remote_rpc remote_session in
-          if List.mem_assoc "host" params then
-            let x = List.assoc "host" params in
-            try
-              List.find
-                (fun (_, h) ->
-                  h.API.host_hostname = x
-                  || h.API.host_name_label = x
-                  || h.API.host_uuid = x
+        let host =
+          let get_host_by_hostname x =
+            let host_matches self =
+              let hostname = remote (Client.Host.get_hostname ~self) in
+              hostname = x
+            in
+            let all_hosts = remote Client.Host.get_all in
+            List.find_opt host_matches all_hosts
+          in
+          let get_host_with x =
+            let ref =
+              if Uuidm.of_string x |> Option.is_some then
+                remote (Client.Host.get_by_uuid ~uuid:x)
+              else
+                let hosts = remote (Client.Host.get_by_name_label ~label:x) in
+                Option.fold ~none:Ref.null ~some:Fun.id (List.nth_opt hosts 0)
+            in
+            if ref <> Ref.null then
+              Some ref
+            else
+              get_host_by_hostname x
+          in
+
+          let search, fail_msg =
+            match List.assoc_opt "host" params with
+            | Some x ->
+                (get_host_with x, Printf.sprintf "Failed to find host: %s" x)
+            | None ->
+                ( List.nth_opt (remote Client.Host.get_all) 0
```

Comment:
This is just `List.hd`?

Reply:
I guess it's like `List.hd_opt`, but that doesn't exist.

Reply:
It's a list.hd that returns an option instead of raising an exception